src/drivers/intel/ssp/mn.c

```diff
@@ -122,9 +122,6 @@ static inline int setup_initial_mclk_source(uint32_t mclk_id,
 
 	/* searching the smallest possible mclk source */
 	for (i = MAX_SSP_FREQ_INDEX; i >= 0; i--) {
-		if (mclk_rate > ssp_freq[i].freq)
-			break;
```

Comment:
I think that was intentional. Removing this case will result in the MCLK rate being an exact integer division of the SSP clock, which completely defeats the purpose of the M/N dividers?

Reply:
> I think that was intentional. Removing this case will result in the MCLK rate being an exact integer division of the SSP clock, which completely defeats the purpose of the M/N dividers?

nope, here we are looking for the most suitable mclk source and the divisible (src_freq % clk_rate == 0) source will be the preferred one, non-1 M/N dividers is not our intention.

We should not stop the searching here at 'mclk_rate > ssp_freq[i].freq', imagine this case:
we have ssp_freq[] = {19.2M, 24.576M, 96M}, and we are asking for 'mclk_rate = 24.576M', without my change, we will stop at 'i = 0' and lose the chance to use 24.576M source and non-1 M/N will be used (which is actually not necessary).

Reply:
You are maybe fixing an issue by adding a new one. that's not cool.

Reply:
> You are maybe fixing an issue by adding a new one. that's not cool.

Adding what new issue?

Reply:
@keyonjie I think that @plbossart means we are enabling M/N when we dont need to in the case where we have an exact integer division i.e. we could use the SSP MDIV bits here for an integer divide to save power.

Reply:
@lgirdwood @plbossart let me clarify things here:
1. for mclk, there is actually no M/N divider concepts, mclk output = mclk source / M_DIV Ratio, see the register definition of MN_MDIVR for detail. Translate it to M/N concept, the M is always 1, and N = 2, 3, ..., 8.
2. the reason why I am removing 'if (mclk_rate > ssp_freq[i].freq)' check here is that the check here is totally logic wrong, the original check will return fail and report "MCLK xxx, no valid source" when we actually have higher frequency mclk source to be used, the higher frequency one is not even iterated with it...

Reply:
@keyonjie can you remind me which cases are currently broken? I don't think we use an MCLK of 24.576MHz in any real product. on APL the mclk does not go out on the connector.

````
sof-apl-asrc-pcm512x.m4:        SSP_CONFIG(I2S, SSP_CLOCK(mclk, 24576000, codec_mclk_in),
sof-apl-asrc-wm8804.m4: SSP_CONFIG(I2S, SSP_CLOCK(mclk, 24576000, codec_mclk_in),
sof-apl-demux-pcm512x.m4:       SSP_CONFIG(I2S, SSP_CLOCK(mclk, 24576000, codec_mclk_in),
sof-apl-nocodec.m4:        SSP_CONFIG(I2S, SSP_CLOCK(mclk, 24576000, codec_mclk_in),
sof-apl-nocodec.m4:        SSP_CONFIG(I2S, SSP_CLOCK(mclk, 24576000, codec_mclk_in),
sof-apl-nocodec.m4:#       SSP_CONFIG(I2S, SSP_CLOCK(mclk, 24576000, codec_mclk_in),
sof-apl-nocodec.m4:        SSP_CONFIG(I2S, SSP_CLOCK(mclk, 24576000, codec_mclk_in),
sof-apl-nocodec.m4:#       SSP_CONFIG(I2S, SSP_CLOCK(mclk, 24576000, codec_mclk_in),
sof-apl-nocodec.m4:        SSP_CONFIG(I2S, SSP_CLOCK(mclk, 24576000, codec_mclk_in),
sof-apl-pcm512x-nohdmi.m4:      SSP_CONFIG(I2S, SSP_CLOCK(mclk, 24576000, codec_mclk_in),
sof-apl-pcm512x.m4:     SSP_CONFIG(I2S, SSP_CLOCK(mclk, 24576000, codec_mclk_in),
sof-apl-pcm512x.m4:     SSP_CONFIG(I2S, SSP_CLOCK(mclk, 24576000, codec_mclk_in),
sof-apl-rt298.m4:       SSP_CONFIG(I2S, SSP_CLOCK(mclk, 24576000, codec_mclk_in),
sof-apl-src-pcm512x.m4: SSP_CONFIG(I2S, SSP_CLOCK(mclk, 24576000, codec_mclk_in),
sof-apl-tdf8532.m4:        SSP_CONFIG(I2S, SSP_CLOCK(mclk, 24576000, codec_mclk_in),
sof-apl-tdf8532.m4:        SSP_CONFIG(I2S, SSP_CLOCK(mclk, 24576000, codec_mclk_in),
sof-apl-tdf8532.m4:        SSP_CONFIG(DSP_B, SSP_CLOCK(mclk, 24576000, codec_mclk_in),
sof-apl-tdf8532.m4:        SSP_CONFIG(I2S, SSP_CLOCK(mclk, 24576000, codec_mclk_in),
sof-apl-tdf8532.m4:        SSP_CONFIG(DSP_B, SSP_CLOCK(mclk, 24576000, codec_mclk_in),
sof-apl-tdf8532.m4:        SSP_CONFIG(I2S, SSP_CLOCK(mclk, 24576000, codec_mclk_in),
sof-apl-wm8804.m4:      SSP_CONFIG(I2S, SSP_CLOCK(mclk, 24576000, codec_mclk_in),
````

I would also recommend you split fixes from functionality change, I find this series complicated to review.

Reply:
@plbossart as you found, the finding for 24.576MHz mclk source here fail for those .m4 files you listed above, though they are not connected out actually, the logic fix here at least no harm. There is no mclk broken reported up to now I think, but let's fix the logic wrong here before it become to a real problem to us.

Yes, let me split those MCLK fixes and talk about them there, thanks for your suggestion.

Reply:
@plbossart the MCLK fixes are split and put here: https://github.com/thesofproject/sof/pull/3672